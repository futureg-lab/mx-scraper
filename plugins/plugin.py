# TODO:
# 1. reqwest wrapper
# 2. afQuery wrapper


import random


def roulette():
    return random.random() > 0.5
