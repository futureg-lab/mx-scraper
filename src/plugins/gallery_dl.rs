use std::{path::PathBuf, process::Command, str::FromStr};

use crate::{
    core::utils::{self, extract_filename},
    schemas::{
        book::{
            Author, Book, Chapter, Metadata, Page, PluginOption, SearchOption, Tag, TitleAlias,
        },
        default_on_null, deserialize_coerce_num_as_opt_string, deserialize_singleton,
    },
};
use anyhow::{Context, Ok};
use indexmap::IndexSet;
use serde::{Deserialize, Serialize};
use url::Url;

use super::MXPlugin;

#[derive(Debug)]
pub struct GalleryDLPlugin {
    pub name: String,
    pub bin: PathBuf,
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Gallery {
    #[serde(default, alias = "manga")]
    title: Option<String>,

    #[serde(default, deserialize_with = "deserialize_singleton")]
    title_aliases: Vec<String>,

    #[serde(
        default,
        alias = "manga_id",
        deserialize_with = "deserialize_coerce_num_as_opt_string"
    )]
    gallery_id: Option<String>,

    #[serde(default, deserialize_with = "default_on_null")]
    user: serde_json::Value,

    #[serde(
        default,
        alias = "artist",
        alias = "authors",
        alias = "author",
        deserialize_with = "deserialize_singleton"
    )]
    artists: Vec<serde_json::Value>,

    #[serde(default, deserialize_with = "default_on_null")]
    group: Vec<String>,

    #[serde(default, deserialize_with = "default_on_null")]
    category: String,

    #[serde(default, deserialize_with = "default_on_null")]
    characters: Vec<String>,

    #[serde(default, deserialize_with = "default_on_null")]
    language: String,

    #[serde(default, deserialize_with = "default_on_null")]
    parody: Vec<String>,

    #[serde(default, deserialize_with = "default_on_null")]
    subcategory: String,

    #[serde(default, deserialize_with = "default_on_null")]
    tags: Vec<String>,

    #[serde(default, deserialize_with = "default_on_null")]
    hashtags: Vec<String>,
}

#[derive(Debug, Serialize, Clone, Deserialize, Default)]
pub struct GalleryPage {
    filename: Option<String>,
    extension: Option<String>,
    #[serde(default)]
    title: String,
    #[serde(default, deserialize_with = "default_on_null")]
    tags: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct FirstGalleryEntry(i32, Gallery);

#[derive(Debug, Serialize, Deserialize)]
pub struct UrlGalleryEntry(i32, String, serde_json::Value);

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum GalleryItem {
    TwoElementTuple(FirstGalleryEntry),
    ThreeElementTuple(UrlGalleryEntry),
}

impl GalleryDLPlugin {
    pub fn new() -> Self {
        Self {
            name: String::from("gallery-dl"),
            bin: PathBuf::from("gallery-dl"),
        }
    }
}

impl MXPlugin for GalleryDLPlugin {
    async fn init(&mut self) -> anyhow::Result<()> {
        Ok(())
    }

    async fn destroy(&mut self) -> anyhow::Result<()> {
        Ok(())
    }

    fn configure(&mut self, _option: PluginOption) -> anyhow::Result<()> {
        Ok(())
    }

    async fn get_book(&self, term: String) -> anyhow::Result<Book> {
        let mut command = Command::new(&self.bin);
        let output = command.arg(&term).arg("--dump-json").output()?;

        if !output.status.success() {
            let lines = &[
                format!("stdout: {}", String::from_utf8_lossy(&output.stdout)),
                format!("stderr: {}", String::from_utf8_lossy(&output.stderr)),
            ];
            anyhow::bail!("{}", lines.join("\n").to_string());
        }

        let items: Vec<GalleryItem> = serde_json::from_slice(&output.stdout)
            .with_context(|| format!("Parse result of '{}'", print_command_invocation(&command)))?;

        generate_book(term, items)
    }

    async fn search(&self, _term: String, _option: SearchOption) -> anyhow::Result<Vec<Book>> {
        unimplemented!()
    }

    async fn is_supported(&self, term: String) -> anyhow::Result<bool> {
        let mut command = Command::new(&self.bin);
        let output = command.arg(&term).arg("--extractor-info").output()?;
        Ok(output.status.success())
    }
}

pub fn generate_book(term: String, items: Vec<GalleryItem>) -> anyhow::Result<Book> {
    let placeholder_title = utils::resume_text(&term, Some(20)).to_string();
    let mut title = placeholder_title.trim().to_owned();
    let mut title_aliases = IndexSet::new();
    let mut gallery_id = title.clone();

    let mut tags: IndexSet<String> = IndexSet::new();
    let mut authors: IndexSet<String> = IndexSet::new();
    let mut metadata: Vec<Metadata> = vec![];

    let mut pages = vec![];
    for (p, item) in items.into_iter().enumerate() {
        match item {
            GalleryItem::TwoElementTuple(FirstGalleryEntry(_len, gl)) => {
                title = gl.title.unwrap_or_else(|| term.clone());
                gallery_id = if gl.gallery_id.is_none() {
                    term.clone()
                } else {
                    gl.gallery_id.unwrap_or_default()
                };

                title_aliases.extend(gl.title_aliases);

                authors.extend(
                    gl.artists
                        .iter()
                        .map(|v| get_if_not_str_rec("name", v))
                        .collect::<anyhow::Result<Vec<_>>>()?,
                );
                authors.extend(gl.group);

                tags.extend(gl.hashtags);
                tags.extend(gl.tags);
                tags.extend(gl.characters);
                tags.extend(gl.parody);
                tags.insert(gl.category);
                tags.insert(gl.language);
            }
            GalleryItem::ThreeElementTuple(UrlGalleryEntry(_len, url, meta)) => {
                assert!(p >= 1, "First image populated");

                let gpage = serde_json::from_value::<GalleryPage>(meta.clone())?;

                let canon_filename = gpage.filename.map(|f| match gpage.extension {
                    None => format!("{f}.jpg"),
                    Some(ext) => format!("{f}.{ext}"),
                });

                let filename = canon_filename.unwrap_or_else(|| {
                    extract_filename(&Url::from_str(&url).unwrap()).unwrap_or_else(|| p.to_string())
                });

                pages.push(Page {
                    title: gpage.title,
                    url,
                    number: p as u32,
                    filename,
                    ..Default::default()
                });

                tags.extend(gpage.tags);

                metadata.push(Metadata {
                    label: format!("Page {p}"),
                    content: meta,
                });
            }
        }
    }

    let book = Book {
        title: title.clone(),
        url: term.clone(),
        title_aliases: title_aliases
            .into_iter()
            .map(|title| TitleAlias {
                title,
                ..Default::default()
            })
            .collect(),
        source_id: gallery_id,
        description: "".to_string(),
        chapters: vec![Chapter {
            title,
            url: term.clone(),
            number: 1,
            pages,
            description: "".to_string(),
        }],
        authors: authors
            .into_iter()
            .filter_map(|name| {
                if name.is_empty() {
                    None
                } else {
                    Some(Author {
                        name,
                        ..Default::default()
                    })
                }
            })
            .collect(),
        tags: tags
            .into_iter()
            .filter_map(|name| {
                if name.is_empty() {
                    None
                } else {
                    Some(Tag {
                        name,
                        metadata: vec![],
                    })
                }
            })
            .collect(),
        metadata,
    };

    Ok(book)
}

fn print_command_invocation(command: &Command) -> String {
    let bin = command.get_program().to_string_lossy();

    let args: Vec<String> = command
        .get_args()
        .map(|arg| arg.to_string_lossy().to_string())
        .collect();

    format!("{} {}", bin, args.join(" "))
}

fn get_if_not_str_rec(field: &str, value: &serde_json::Value) -> anyhow::Result<String> {
    match value {
        serde_json::Value::Object(m) => {
            if let Some(v) = m.get(field) {
                return get_if_not_str_rec(field, v);
            }
            anyhow::bail!("Field 'name' not found in {m:?}")
        }
        serde_json::Value::String(s) => Ok(s.clone()),
        _ => anyhow::bail!("Could not get name from {value:?}"),
    }
}
