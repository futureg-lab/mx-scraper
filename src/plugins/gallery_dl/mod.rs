use std::{
    path::{Path, PathBuf},
    process::Command,
    str::FromStr,
};

use crate::{
    core::utils::{self, extract_filename},
    schemas::book::{Author, Book, Chapter, Metadata, Page, SearchOption, Tag, TitleAlias},
    GLOBAL_CONFIG,
};
use anyhow::Ok;
use indexmap::IndexSet;
use schema::{FirstGalleryEntry, GalleryItem, GalleryPage, UrlGalleryEntry};
use url::Url;

use super::MXPlugin;
pub mod schema;

#[derive(Debug, Clone)]
pub struct GalleryDLPlugin {
    pub name: String,
    pub bin: PathBuf,
}

impl GalleryDLPlugin {
    pub fn new() -> Self {
        let name = String::from("gallery-dl");
        Self {
            bin: {
                let bin = GLOBAL_CONFIG
                    .read()
                    .unwrap()
                    .request
                    .get(&name)
                    .map(|req| {
                        req.extra_config
                            .clone()
                            .map(|cfg| cfg.get("bin").cloned())
                            .flatten()
                    })
                    .flatten();

                match bin {
                    Some(path) => PathBuf::from(path),
                    None => PathBuf::from("gallery-dl"),
                }
            },

            name,
        }
    }
}

impl MXPlugin for GalleryDLPlugin {
    async fn init(&mut self) -> anyhow::Result<()> {
        Ok(())
    }

    async fn destroy(&mut self) -> anyhow::Result<()> {
        Ok(())
    }

    async fn get_book(&self, term: String) -> anyhow::Result<Book> {
        let mut command = Command::new(&self.bin);
        let output = command.arg(&term).arg("--dump-json").output()?;

        if !output.status.success() {
            let lines = &[
                format!("stdout: {}", String::from_utf8_lossy(&output.stdout)),
                format!("stderr: {}", String::from_utf8_lossy(&output.stderr)),
            ];
            anyhow::bail!("{}", lines.join("\n").to_string());
        }

        let items: Vec<GalleryItem> = serde_json::from_slice(&output.stdout).map_err(|e| {
            anyhow::anyhow!(
                "Parse result of '{}': {e}",
                print_command_invocation(&command)
            )
        })?;

        generate_book(term, items)
    }

    async fn search(&self, _term: String, _option: SearchOption) -> anyhow::Result<Vec<Book>> {
        unimplemented!()
    }

    async fn is_supported(&self, term: String) -> anyhow::Result<bool> {
        let mut command = Command::new(&self.bin);
        let output = command.arg(&term).arg("--extractor-info").output()?;
        Ok(output.status.success())
    }

    fn download_url(&self, dest: &Path, url: &Url) -> Option<anyhow::Result<()>> {
        let body = || -> anyhow::Result<()> {
            let mut command = Command::new(&self.bin);
            let parent = dest.parent().unwrap();
            let filename = dest.file_name().unwrap();
            let tmp_dest = parent.join(format!("{}_temp", filename.to_string_lossy()));

            let output = command
                .arg("--directory")
                .arg(&tmp_dest)
                .arg(url.to_string())
                .output()?;

            if output.status.success() {
                // dl expects dest to be a directory, which is not the case
                // actual/path/page.jpg_temp/custom_name_by_dl.jpg
                // => actual/path/page.jpg
                for entry in std::fs::read_dir(&tmp_dest)? {
                    std::fs::rename(&tmp_dest.join(entry?.file_name()), &dest)?;
                    std::fs::remove_dir_all(&tmp_dest)?;
                    break;
                }
            }

            Ok(())
        };
        Some(body())
    }
}

pub fn generate_book(term: String, items: Vec<GalleryItem>) -> anyhow::Result<Book> {
    let placeholder_title = utils::resume_text(&term, Some(20)).to_string();
    let mut title = placeholder_title.trim().to_owned();
    let mut title_aliases = IndexSet::new();
    let mut gallery_id = title.clone();

    let mut tags: IndexSet<String> = IndexSet::new();
    let mut authors: IndexSet<String> = IndexSet::new();

    let mut pages = vec![];
    for (p, item) in items.iter().enumerate() {
        match item {
            GalleryItem::TwoElementTuple(FirstGalleryEntry(_len, gl)) => {
                title = gl.get_title_or_default(term.clone());
                gallery_id = gl.get_id_or_default(term.clone());

                title_aliases.extend(gl.get_title_aliases());
                tags.extend(gl.get_tags());
                authors.extend(gl.get_authors()?);
            }
            GalleryItem::ThreeElementTuple(UrlGalleryEntry(_len, url, meta)) => {
                let gpage = serde_json::from_value::<GalleryPage>(meta.clone())?;

                let filename = gpage.get_filename().unwrap_or_else(|| {
                    extract_filename(&Url::from_str(&url).unwrap()).unwrap_or_else(|| p.to_string())
                });

                let page_meta = meta.as_object().map(|o| {
                    if !o.is_empty() {
                        vec![Metadata {
                            label: filename.clone(),
                            content: meta.clone(),
                        }]
                    } else {
                        vec![]
                    }
                });

                tags.extend(gpage.get_explicit_tags());

                pages.push(Page {
                    title: gpage.title,
                    url: url.clone(),
                    number: p as u32,
                    filename,
                    metadata: page_meta.unwrap_or_else(|| vec![]),
                    ..Default::default()
                });
            }
        }
    }

    gallery_id = utils::set_if_empty(
        gallery_id,
        utils::compute_query_signature(&term, "gallery-dl"),
    );

    let book = Book {
        title: title.clone(),
        url: term.clone(),
        title_aliases: title_aliases
            .into_iter()
            .map(|title| TitleAlias {
                title,
                ..Default::default()
            })
            .collect(),
        source_id: gallery_id,
        description: "".to_string(),
        chapters: vec![Chapter {
            title,
            url: term.clone(),
            number: 1,
            pages,
            description: "".to_string(),
            ..Default::default()
        }],
        authors: authors
            .into_iter()
            .filter_map(|name| {
                if name.is_empty() {
                    None
                } else {
                    Some(Author {
                        name,
                        ..Default::default()
                    })
                }
            })
            .collect(),
        tags: tags
            .into_iter()
            .filter_map(|name| {
                if name.is_empty() {
                    None
                } else {
                    Some(Tag {
                        name,
                        metadata: vec![],
                    })
                }
            })
            .collect(),
        metadata: vec![],
    };

    Ok(book)
}

fn print_command_invocation(command: &Command) -> String {
    let bin = command.get_program().to_string_lossy();

    let args: Vec<String> = command
        .get_args()
        .map(|arg| arg.to_string_lossy().to_string())
        .collect();

    format!("{} {}", bin, args.join(" "))
}
