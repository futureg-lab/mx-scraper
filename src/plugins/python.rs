use std::{
    path::{Path, PathBuf},
    str::FromStr,
};

use crate::{
    core::http::{self},
    schemas::book::{Book, PluginOption, SearchOption},
    GLOBAL_CONFIG,
};
use anyhow::{bail, Context, Ok};
use pyo3::{exceptions::PyException, prelude::*, types::PyBytes};
use serde_pyobject::{from_pyobject, to_pyobject};
use url::Url;

use super::MXPlugin;

#[derive(Debug)]
pub struct PythonPlugin {
    pub name: String,
    pub workdir: Option<PathBuf>,
}

impl MXPlugin for PythonPlugin {
    async fn init(&mut self) -> anyhow::Result<()> {
        if self.name.split_whitespace().count() > 1 {
            bail!(
                "Invalid {:?}: plugin name cannot contain whitespaces",
                self.name
            )
        }

        if self.workdir.is_none() {
            self.workdir = Some(Path::new("./plugins").canonicalize()?);
        }

        std::env::set_var("PYTHONPATH", self.workdir.clone().unwrap());

        Ok(())
    }

    async fn destroy(&mut self) -> anyhow::Result<()> {
        Ok(())
    }

    fn configure(&mut self, _option: PluginOption) -> anyhow::Result<()> {
        Ok(())
    }

    async fn get_book(&self, term: String) -> anyhow::Result<Book> {
        pyo3::prepare_freethreaded_python();
        Python::with_gil(|py| {
            let name: &str = self.name.as_ref();
            let plugin = py.import_bound(name)?;
            let mx_request = MxRequest;

            if plugin.hasattr("mx_get_urls")? {
                let res = plugin
                    .call_method1("mx_get_urls", (term.clone(), mx_request))
                    .with_context(|| {
                        format!("{}.mx_get_urls(term = {term:?}, ..)", self.name.clone())
                    })?;
                Book::from_raw_urls(from_pyobject(res)?)
            } else if plugin.hasattr("mx_get_book")? {
                match plugin.call_method1("mx_get_book", (term.clone(), mx_request)) {
                    Err(py_err) => {
                        bail!(
                            "{}.mx_get_book(term = {term:?}, ..): {py_err}",
                            self.name.clone()
                        )
                    }
                    other => {
                        let other = other?;
                        from_pyobject(other.clone())
                            .with_context(|| format!("Deserializing {:?}", other))
                    }
                }
            } else {
                bail!("Invalid could not find mx_get_urls(term) or mx_get_book(term)",)
            }
        })
    }

    async fn search(&self, _term: String, _option: SearchOption) -> anyhow::Result<Vec<Book>> {
        unimplemented!()
    }

    async fn is_supported(&self, term: String) -> anyhow::Result<bool> {
        pyo3::prepare_freethreaded_python();
        Python::with_gil(|py| {
            let name: &str = self.name.as_ref();
            let plugin = py.import_bound(name)?;
            let res = plugin
                .call_method1("mx_is_supported", (term.clone(),))
                .with_context(|| format!("Calling mx_is_supported with term {term:?}"))?;
            res.extract()
                .with_context(|| format!("mx_is_supported returned {res:?} but bool was expected",))
        })
    }
}

#[pyclass]
#[derive(Debug)]
pub struct MxRequest;

macro_rules! can_throw_exception {
    ($e: expr) => {
        $e.map_err(|e| PyException::new_err(e.to_string()))?
    };
}

#[pymethods]
impl MxRequest {
    #[pyo3(signature = (url, context=None))]
    fn fetch(
        &self,
        py: Python,
        url: String,
        context: Option<Bound<PyAny>>,
    ) -> PyResult<Py<PyBytes>> {
        let url = can_throw_exception!(Url::from_str(&url));

        let bytes = can_throw_exception!(match context {
            Some(py_context) => {
                let context = from_pyobject(py_context)?;
                http::fetch_with_context(url, context)
            }
            None => http::fetch(url),
        });

        let bytes = PyBytes::new_bound(py, bytes.as_ref()).unbind();
        std::result::Result::Ok(bytes)
    }

    #[getter]
    pub fn context(&self, py: Python) -> Py<PyAny> {
        let context = {
            let config = GLOBAL_CONFIG.lock().unwrap();
            config.gen_fetch_context()
        };
        to_pyobject(py, &context).unwrap().unbind()
    }
}
