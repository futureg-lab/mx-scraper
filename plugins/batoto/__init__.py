from typing import Dict, Any
from plugin import MxRequest
from batoto.utils import fetch_book

PREFIX = "to:"
DOMAINS = ["https://mto.to", "https://xbato.com"]


def pick_base(term: str):
    for base in DOMAINS:
        if term.startswith(base):
            return base
    return None


def mx_is_supported(term: str):
    return term.startswith(PREFIX) or pick_base(term.removeprefix(PREFIX)) is not None


def mx_get_book(term: str, req: MxRequest) -> Dict[str, Any]:
    term = term.removeprefix(PREFIX)
    base = pick_base(term)
    return fetch_book(term, base, req)
