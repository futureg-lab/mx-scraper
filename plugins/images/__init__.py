from typing import Dict, List, Tuple, Any
from urllib.parse import urljoin

from bs4 import BeautifulSoup
from plugin import MxRequest

"""
This will collect all images from an url.
Requires bs4 installed:

pip install beautifulsoup4
"""

PREFIX = "img:"


def mx_is_supported(term: str) -> Dict[str, Any]:
    return term.startswith(PREFIX)


def mx_get_urls(term: str, req: MxRequest) -> Tuple[str, str, List[str], List[str]]:
    url = term.removeprefix(PREFIX)
    response = req.fetch(url).decode("utf-8")
    soup = BeautifulSoup(response, "html.parser")
    title = soup.title.string if soup.title else "No title found"
    img_sources = [
        urljoin(url, img["src"]) for img in soup.find_all("img") if "src" in img.attrs
    ]

    return {
        "title": title,
        "url_source": url,
        "urls": img_sources,
        "tags": [],
    }
