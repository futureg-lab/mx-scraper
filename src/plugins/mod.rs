use std::time::Duration;

use anyhow::Context;
use python::PythonPlugin;

use crate::{
    schemas::book::{Book, PluginOption, SearchOption},
    GLOBAL_CONFIG,
};
pub mod python;

pub trait MXPlugin {
    async fn init(&mut self) -> anyhow::Result<()>;
    async fn destroy(&mut self) -> anyhow::Result<()>;
    fn configure(&mut self, option: PluginOption) -> anyhow::Result<()>;
    async fn get_book(&self, query: String) -> anyhow::Result<Book>;
    async fn is_supported(&self, query: String) -> anyhow::Result<bool>;
    async fn search(&self, term: String, option: SearchOption) -> anyhow::Result<Vec<Book>>;
}

#[derive(Debug)]
pub enum PluginImpl {
    Python(PythonPlugin),
    // TODO:
    // Lua(LuaPlugin)
    // OldNXScraper(OldNXScraperPlugin) // full rust
    // GalleryDL(GalleryDLPlugin) // full rust
}

pub struct PluginManager {
    plugins: Vec<PluginImpl>,
}

#[derive(Debug, Clone)]
pub struct FetchResult {
    pub query_term: String,
    pub book: Book,
    pub plugin_name: String,
    pub cached: bool,
}

impl PluginManager {
    pub fn new() -> Self {
        Self { plugins: vec![] }
    }

    /// Fetch a term using the first plugin that supports it
    pub async fn auto_fetch(&self, term: String) -> anyhow::Result<FetchResult> {
        let mut issues = vec![];
        for plugin in &self.plugins {
            match plugin {
                PluginImpl::Python(py) => {
                    if py.is_supported(term.clone()).await? {
                        match self.fetch(term.clone(), py.name.clone()).await {
                            Ok(f) => return Ok(f),
                            Err(e) => {
                                issues.push(format!("  - Plugin {}: {e}", py.name));
                            }
                        }
                    }
                }
            }
        }

        if !issues.is_empty() {
            anyhow::bail!("\n{}", issues.join("\n"))
        } else {
            anyhow::bail!("Cannot auto-detect plugin that supports the term {term:?}\nYou can try with --plugin <PLUGIN_NAME>")
        }
    }

    /// Fetch and bypass term validation
    pub async fn fetch(&self, term: String, plugin_name: String) -> anyhow::Result<FetchResult> {
        for plugin in &self.plugins {
            match plugin {
                PluginImpl::Python(py) => {
                    if py.name.eq(&plugin_name) {
                        let (enable_cache, cache_file_path, delay) = {
                            let config = GLOBAL_CONFIG.read().unwrap();
                            (
                                config.cache.enable,
                                config.get_cache_file_path(&term, &plugin_name),
                                config.delay.clone(),
                            )
                        };

                        let mut cached = false;

                        let book = if enable_cache && cache_file_path.exists() {
                            cached = true;
                            let content = std::fs::read_to_string(&cache_file_path)?;
                            serde_json::from_str(&content).with_context(|| {
                                format!(
                                    "Deserializing cache file for term {term} located at {:?}",
                                    cache_file_path.display()
                                )
                            })?
                        } else {
                            let book = py.get_book(term.clone()).await?;
                            let content = serde_json::to_string_pretty(&book)?;
                            std::fs::write(&cache_file_path, content).with_context(|| {
                                format!("Writing cache {:?}", cache_file_path.display())
                            })?;
                            book
                        };
                        tokio::time::sleep(Duration::from_millis(delay.fetch as u64)).await;

                        return Ok(FetchResult {
                            query_term: term.clone(),
                            book,
                            plugin_name,
                            cached,
                        });
                    }
                }
            }
        }

        anyhow::bail!("No plugin named {plugin_name:?}")
    }

    /// A list of all installed plugins
    pub fn list_plugins(&self) -> Vec<String> {
        self.plugins
            .iter()
            .map(|plugin| match plugin {
                PluginImpl::Python(py) => py.name.clone(),
            })
            .collect()
    }

    /// A list of all installed plugins
    pub fn assert_exists(&self, plugin_name: String) -> anyhow::Result<()> {
        if !self.list_plugins().contains(&plugin_name) {
            anyhow::bail!("Plugin named {plugin_name:?} does not exist")
        }
        Ok(())
    }

    /// Initialize all plugins
    pub async fn init(&mut self) -> anyhow::Result<()> {
        let location = { GLOBAL_CONFIG.read().unwrap().plugins.clone().location };
        self.prepare_folders();

        // Add static plugins then collect dynamic ones
        // For now, we have dynamic ones (python)
        // static plugins can be added directly in the Vec below

        let mut plugins = vec![];
        let plug_dir = location.canonicalize()?;

        // +-- plugin_location
        //   +- foo
        //      + __init__.py
        //   +- bar
        //      + __init__.py

        for entry in plug_dir.read_dir()? {
            let entry = entry?;
            let plugin_name = entry.file_name().to_string_lossy().to_string();
            let workdir = plug_dir.join(plugin_name.clone()).to_path_buf();
            let actual_plugin = workdir.join("__init__.py");
            if actual_plugin.exists() {
                // println!(" + Discovered {plugin_name}");
                let python = PythonPlugin {
                    name: plugin_name.clone(),
                    workdir: None,
                };
                plugins.push(PluginImpl::Python(python));
            }
        }

        // init all
        for plugin in plugins.iter_mut() {
            match plugin {
                PluginImpl::Python(py) => py.init().await?,
            }
        }

        self.plugins = plugins;
        Ok(())
    }

    fn prepare_folders(&self) {
        let (cache_folder, download, temp, metadata, plugins) = {
            let config = GLOBAL_CONFIG.read().unwrap();
            (
                config.cache.folder.clone(),
                config.download_folder.download.clone(),
                config.download_folder.temp.clone(),
                config.download_folder.metadata.clone(),
                config.plugins.location.clone(),
            )
        };

        std::fs::create_dir_all(&cache_folder)
            .with_context(|| format!("Create cache folder {:?}", cache_folder.display()))
            .unwrap();

        std::fs::create_dir_all(&temp)
            .with_context(|| format!("Create temp folder {:?}", temp.display()))
            .unwrap();

        std::fs::create_dir_all(&download)
            .with_context(|| format!("Create download folder {:?}", download.display()))
            .unwrap();

        std::fs::create_dir_all(&metadata)
            .with_context(|| format!("Create metadata folder {:?}", metadata.display()))
            .unwrap();

        std::fs::create_dir_all(&plugins)
            .with_context(|| format!("Create plugin folder {:?}", plugins.display()))
            .unwrap();
    }

    /// Destroy all plugins and free-up ressources
    pub async fn destroy(&mut self) -> anyhow::Result<()> {
        for plugin in self.plugins.iter_mut() {
            match plugin {
                PluginImpl::Python(py) => py.destroy().await?,
            }
        }
        Ok(())
    }
}
