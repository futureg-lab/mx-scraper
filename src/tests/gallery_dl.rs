use insta::assert_debug_snapshot;

use crate::plugins::{
    gallery_dl::{self, Gallery, GalleryDLPlugin, GalleryItem},
    MXPlugin,
};

// nh
#[test]
fn parse_gallery_dl_first_item_example_1() {
    let content =
        std::fs::read_to_string("./src/tests/gallery_dl/first_item_example_1.json").unwrap();
    serde_json::from_str::<Gallery>(&content).unwrap();
}

#[test]
fn parse_gallery_dl_generic_example_1() {
    let content = std::fs::read_to_string("./src/tests/gallery_dl/example_1.json").unwrap();
    serde_json::from_str::<Vec<GalleryItem>>(&content).unwrap();
}

// kem
#[test]
fn parse_gallery_dl_first_item_example_2() {
    let content =
        std::fs::read_to_string("./src/tests/gallery_dl/first_item_example_2.json").unwrap();
    serde_json::from_str::<Gallery>(&content).unwrap();
}

#[test]
fn parse_gallery_dl_generic_example_2() {
    let content = std::fs::read_to_string("./src/tests/gallery_dl/example_2.json").unwrap();
    serde_json::from_str::<Vec<GalleryItem>>(&content).unwrap();
}

// tw
#[test]
fn parse_gallery_dl_first_item_example_3() {
    let content =
        std::fs::read_to_string("./src/tests/gallery_dl/first_item_example_3.json").unwrap();
    let output = serde_json::from_str::<Vec<GalleryItem>>(&content).unwrap();
    let book = gallery_dl::generate_book("http://some.website/a/b/c".to_string(), output).unwrap();
    assert_debug_snapshot!(book);
}

#[tokio::test]
async fn fetch_book() {
    let mut plugin = GalleryDLPlugin::new();

    let term = "https://twitter.com/imigimuru/status/1829913427373953259".to_string();

    plugin.init().await.unwrap();

    if !plugin.is_supported(term.clone()).await.unwrap() {
        panic!("Sauce {term:?} not supported?");
    }

    plugin.get_book(term).await.unwrap();
}
