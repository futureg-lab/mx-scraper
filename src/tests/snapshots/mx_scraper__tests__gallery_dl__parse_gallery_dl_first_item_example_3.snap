---
source: src/tests/gallery_dl.rs
expression: book
---
Book {
    title: "http://some.website/a/b/c",
    title_aliases: [],
    source_id: "http://some.website/a/b/c",
    description: "",
    authors: [
        Author {
            name: "imigimuru",
            description: "",
        },
    ],
    chapters: [
        Chapter {
            title: "http://some.website/a/b/c",
            description: "",
            url: "http://some.website/a/b/c",
            number: 1,
            pages: [
                Page {
                    title: "",
                    url: "https://pbs.twimg.com/media/GWUoj32a4AAwuqN?format=jpg&name=orig",
                    intermediate_link_hint: None,
                    number: 1,
                    filename: "GWUoj32a4AAwuqN.jpg",
                },
            ],
        },
    ],
    tags: [
        Tag {
            name: "マケイン",
            metadata: [],
        },
        Tag {
            name: "twitter",
            metadata: [],
        },
    ],
    metadata: [
        Metadata {
            label: "Page 1",
            content: Object {
                "author": Object {
                    "date": String("2012-02-04 06:50:39"),
                    "description": String("漫画家。『この美術部には問題がある！』連載中。【FGO/概念礼装】【負けヒロインが多すぎる！/イラスト】【リコリス・リコイル/キャラクターデザイン】初画集『fruits』発売中。お仕事はこちら→ayac2006☆http://yahoo.co.jp (☆→@) アーセナルが好き⚽\u{fe0f}"),
                    "favourites_count": Number(20677),
                    "followers_count": Number(349650),
                    "friends_count": Number(1214),
                    "id": Number(482708089),
                    "listed_count": Number(3140),
                    "location": String("兵庫県"),
                    "media_count": Number(1852),
                    "name": String("imigimuru"),
                    "nick": String("いみぎむる"),
                    "profile_banner": String("https://pbs.twimg.com/profile_banners/482708089/1399750342"),
                    "profile_image": String("https://pbs.twimg.com/profile_images/875689427502223360/dl_quOOM.jpg"),
                    "statuses_count": Number(58098),
                    "verified": Bool(false),
                },
                "category": String("twitter"),
                "content": String("「負けヒロインが多すぎる！」第8話ご視聴ありがとうございました🚰\n次回もぜひよろしくお願いします❤\u{fe0f}\u{200d}🩹\n\n#マケイン"),
                "conversation_id": Number(1829913427373953259),
                "count": Number(1),
                "date": String("2024-08-31 16:05:33"),
                "description": Null,
                "extension": String("jpg"),
                "favorite_count": Number(6531),
                "filename": String("GWUoj32a4AAwuqN"),
                "hashtags": Array [
                    String("マケイン"),
                ],
                "height": Number(1270),
                "lang": String("ja"),
                "num": Number(1),
                "quote_count": Number(37),
                "quote_id": Number(0),
                "reply_count": Number(49),
                "reply_id": Number(0),
                "retweet_count": Number(1407),
                "retweet_id": Number(0),
                "sensitive": Bool(false),
                "source": String("Twitter for iPhone"),
                "subcategory": String("tweet"),
                "tweet_id": Number(1829913427373953259),
                "user": Object {
                    "date": String("2012-02-04 06:50:39"),
                    "description": String("漫画家。『この美術部には問題がある！』連載中。【FGO/概念礼装】【負けヒロインが多すぎる！/イラスト】【リコリス・リコイル/キャラクターデザイン】初画集『fruits』発売中。お仕事はこちら→ayac2006☆http://yahoo.co.jp (☆→@) アーセナルが好き⚽\u{fe0f}"),
                    "favourites_count": Number(20677),
                    "followers_count": Number(349650),
                    "friends_count": Number(1214),
                    "id": Number(482708089),
                    "listed_count": Number(3140),
                    "location": String("兵庫県"),
                    "media_count": Number(1852),
                    "name": String("imigimuru"),
                    "nick": String("いみぎむる"),
                    "profile_banner": String("https://pbs.twimg.com/profile_banners/482708089/1399750342"),
                    "profile_image": String("https://pbs.twimg.com/profile_images/875689427502223360/dl_quOOM.jpg"),
                    "statuses_count": Number(58098),
                    "verified": Bool(false),
                },
                "width": Number(1300),
            },
        },
    ],
    url: "http://some.website/a/b/c",
}
